engine.go

```diff
@@ -116,15 +117,16 @@ func (p *PreparedDataCache) UncacheStmt(sessId uint32, query string) {
 
 // Engine is a SQL engine.
 type Engine struct {
-	Analyzer          *analyzer.Analyzer
-	LS                *sql.LockSubsystem
-	ProcessList       sql.ProcessList
-	MemoryManager     *sql.MemoryManager
-	BackgroundThreads *sql.BackgroundThreads
-	IsReadOnly        bool
-	IsServerLocked    bool
-	PreparedDataCache *PreparedDataCache
-	mu                *sync.Mutex
+	Analyzer                *analyzer.Analyzer
+	LS                      *sql.LockSubsystem
+	ProcessList             sql.ProcessList
+	MemoryManager           *sql.MemoryManager
+	BackgroundThreads       *sql.BackgroundThreads
+	IsReadOnly              bool
+	IsServerLocked          bool
+	PreparedDataCache       *PreparedDataCache
+	BinlogReplicaController binlogreplication.BinlogReplicaController
```

Comment:
This almost seems misplaced. The only reason this needs to be here is to be assigned to certain plan nodes, which could also happen via an analysis rule, and the analyzer could be constructed with a controller instead. I do agree it seems like a replication controller *should* be an engine-level concern, but as it's written now it doesn't seem to have a place here. Is there future functionality that makes you think this needs to be managed by the engine?

Reply:
I think you're right – applying the replica controller to the statements that need it is cleaner as an analyzer rule. I just pushed a new commit that makes that switch.